Improve first person camera in example

janhohenheim opened the pull request:
# Objective

- I've seen quite a few people on discord copy-paste the camera code of the first-person example and then run into problems with the pitch.
- ~~Additionally, the code is framerate-dependent.~~ it's not, see comment in PR

## Solution

- Make the code good enough to be copy-pasteable 
	- ~~Use `dt` to make the code framerate-independent~~ Add comment explaining why we don't use `dt`
	- Clamp the pitch
	- Move the camera sensitivity into a component for better configurability

## Testing

Didn't run the example again, but the code is straight from another project I have open, so I'm not worried.

NiseVoid commented:
Multiplying mouse motion by delta time actually has the opposite effect. The mouse motion already matches the amount you moved it during the frame, so multipling it by delta time makes your mouse movement FPS-dependant, with a higher FPS meaning slower movement. However, if the input was a gamepad stick it would need to be multiplied by delta, which might be worth including in the example for that reason. Ideally we would of course upstream LWIM and have it handle those differences instead.

I do like the limit for the angle, tho the comment could be better. Not only is rotating too far not particularly useful, it also makes the other axis function like it's reversed.

alice-i-cecile commented on examples/camera/first_person_view_model.rs:
```diff
@@ -219,18 +221,33 @@ fn spawn_text(mut commands: Commands) {
     let mut transform = player.single_mut();
     let delta = accumulated_mouse_motion.delta;
+    let dt = time.delta_seconds();
+    // The factors are just arbitrary mouse sensitivity values.
```
Pull these out into constants please :)

janhohenheim replied:
I disagree. Pulling them into `MOUSE_SENSITIVITY_X` and `MOUSE_SENSITIVITY_Y`
- makes them less nicely grouped than now, and
- is semantically wrong. `PI` is a constant, an aspect ratio is a constant. Mouse sensitivity is a value that should be a configurable variable for accessibility.

As a compromise, I could extract them into a component if you want. Or add a component mentioning that this should be configurable.

alice-i-cecile replied:
Component is nice: it makes it easy for the values to be twiddled via inspectors.

janhohenheim replied:
[`3e372d1` (#15109)](https://github.com/bevyengine/bevy/pull/15109/commits/3e372d1362bdd8a519dde4bcabbc4983122225b6)

alice-i-cecile commented:
@richchurcher, if you're interested in helping out by reviewing, I'd welcome your opinions here :) Bevy uses community reviews to vet when things are ready to merge, so please leave an approval when you think this work is ready.

janhohenheim commented:
@NiseVoid thanks for pointing out that `dt` makes the mouse actually frame dependent in this case. I didn't consider that before, but it totally makes sense. Time to change all the code that works like this in my other projects :D 
Addressed all review comments.

alice-i-cecile reviewed (commented):
Sticking data on the `Player` component like this feels messy, but it *does* keep the example simpler. I'd prefer a dedicated `CameraSensitivity` component.

janhohenheim commented:
@alice-i-cecile yeah I was considering that as well and used the `Player` component in fear of bloat. But you're right, showing good ECS patterns is probably worth it here. Changed it :)

chompaa commented on examples/camera/first_person_view_model.rs:
```diff
@@ -220,25 +239,46 @@ fn spawn_text(mut commands: Commands) {
+        // The camera has no way of knowing what direction was "forward" before landing in that extreme position,
+        // so the direction picked will for all intents and purposes be arbitrary.
+        // Another issue is that for mathematical reasons, the yaw will effectively be flipped when the pitch is at the extremes.
+        // To not run into these issues, we clamp the pitch to a safe
```
```suggestion
        // To not run into these issues, we clamp the pitch to a safe range.
```

chompaa commented on examples/camera/first_person_view_model.rs:
```diff
@@ -220,25 +239,46 @@ fn spawn_text(mut commands: Commands) {
+        let yaw = yaw + delta_yaw;
+
+        // If the pitch was ±¹⁄₂ π, the camera would look straight up or down.
+        // When the user want to move the camera back to the horizon, which way should the camera face?
```
```suggestion
        // When the user wants to move the camera back to the horizon, which way should the camera face?
```

richchurcher commented on examples/camera/first_person_view_model.rs:
```diff
@@ -220,25 +239,46 @@ fn spawn_text(mut commands: Commands) {
+        // The camera has no way of knowing what direction was "forward" before landing in that extreme position,
+        // so the direction picked will for all intents and purposes be arbitrary.
+        // Another issue is that for mathematical reasons, the yaw will effectively be flipped when the pitch is at the extremes.
+        // To not run into these issues, we clamp the pitch to a safe range.
```
```suggestion
        // To not run into these issues, we clamp the pitch to a safe range. (FRAC_PI_2 is just a constant representing half the value of π).
```
Feel free to ignore this, but my first glimpse of this value gave me a bit of a WTF moment :laughing:

janhohenheim replied:
Good point :) I changed the wording a bit: [`2bc51d1` (#15109)](https://github.com/bevyengine/bevy/pull/15109/commits/2bc51d11968cd78ead6b4c82119ed18ab8f05e10)

richchurcher reviewed (approved):
All looks very familiar!
